### ISSUE TITLE
Implement a global config file support

### ISSUE THREAD

**Issue by KOLANICH** [pypa/build#146]
It is very inconvenient to provide the same options to software every time. The common solution to that is storing them in a config file. It would be nice to make this tool to check some paths for a presence of config files (i.e. the similar way pip does), and if a valid config file is present there, apply the prefs from it.

**Comment by gaborbernat** [comment]
And what would you configure in this file?

**Comment by FFY00** [comment]
@gaborbernat please do not close right away, give people time to respond :blush:

I imagine it could make a tiny bit of sense with `config_settings`, but I am *very* skeptical of that.

**Comment by gaborbernat** [comment]
I would have left it open if it had any description attached to it. But, this does not, any in my experience for such tickets if people don't respond immediately they're unlikely they'll do. And we can always reopen a ticket once they provided enough information to make it a valid one. I really wish Github would not allow people to open tickets that have no description.

**Comment by FFY00** [comment]
My experience is different. I don't think it's very reasonable to close because someone hasn't responded within 1h :confused:

I usually wait for ~1week before closing, and when I close I try to always say that they are free to reopen. You can shorten it to a few days if you must, it does not cost us any money to keep the tickets open :stuck_out_tongue:

I think that makes for a slightly better upstream interaction and avoids people from getting wrong impressions.

**Comment by KOLANICH** [comment]
>And what would you configure in this file?

`--no-isolation`

**Comment by gaborbernat** [comment]
I'm not to keen personally on this because ``--no-isolation`` assumes the users took care of setting up the environment, this is for the advanced user. The configuration would really hurt the generic user because once it sets it all uses in all environments will suddenly use this automatically (even if e.g. tox/virtualenv relies on the fact of isolation is on). I can see a lot of not experienced people tripping over such functionality... 🤷 As long as we have a single flag IMHO the complexity added does not yet warrant adding a configuration file complexity.

**Comment by KOLANICH** [comment]
It is an opt-in functionality. Someone still has to sudo into root and edit the config file to enable it by default.

>As long as we have a single flag

It is inacceptable to burden me to enter this flag every time. I can create a bash alias, but it just an dirty hack, which won't work in wide range of cases, such as third-party software using `build` API.

**Comment by pganssle** [comment]
I don't see why you would want this. The `--no-isolation` flag is really for distro packagers or other people explicitly creating their own isolated build environments like a chroot.

At the very least it would be something you want to configure per project (like if your project has some specific non-python dependencies and you build it in a docker environment).

You should emphatically not be globally using `--no-isolation `, which would be bad for the python ecosystem in general. If there were a global config file, I would recommend not even supporting `--no-isolation` in it, so I'm pretty strongly -1 on this idea.

**Comment by FFY00** [comment]
Yeah, I also do not think a config file is a place for the `--no-isolation` flag.

We could maybe consider allowing this to be set via an environment variable. @gaborbernat @pganssle, what do you think? I am not sure myself.

> It is inacceptable to burden me to enter this flag every time. I can create a bash alias, but it just an dirty hack, which won't work in wide range of cases, such as third-party software (which can even need source code modifications to ovecome the default, if there is no another mechanism for it) using `build` API. Another alternative is to just use a fork of `build`, but I don't like it.

Well, you can just `-n`, which shouldn't be *that* much of a burden.

**Comment by KOLANICH** [comment]
>I don't see why you would want this.

1. I have some own packages (in fact, setuptools plugins) out of pypi and in quite experimental stage installed with pip `-e` flag, related to building of packages. I need changes in them be captured automatically.
2. I don't trust pypi and try stay away of it. So I deactivated automatic installation of stuff from pypi in pip.
3. IMHO all this build isolation brings no real benefits and has significant drawbacks, like the overhead and non-using bleeding edge unreleased software installed on my machine.

>You should emphatically not be globally using --no-isolation , which would be bad for the python ecosystem in general.

IMHO it is using virtual environments where they are not really needed harmful for python ecosystem in general, since they conceal issues, cause overhead, and promote the mindset that concewling issues is getting rid of them and that concealing issues at the cost of overhead is the right way to do the things. It is an easy way, but not right one. The reality is that cloaking issues related to new versions of software is just locking own packages, and users of them, to old versions of their dependencies, and any particular versions of any software by their nature have no future, so locking anything to any particular version of software means depriving it from future and future from it. So, my motto in package management is "bleeding edge everywhere everytime, where feasible".

**Comment by FFY00** [comment]
That is a reasonable use-case. I think the forceful build isolation in PEP 517 is very opinionated. That is great for users who don't know what they want, but hurts users who do know. That is the whole reason I created this project, I needed a PEP 517 builder without isolation.

> It is easy to forget to enter that.

Sure. I usually my shell lets me autocomple to the last matching command automatically, so I am not usually hit by this kind of issues.

> And it doesn't solve the issue with packages using `build` via API.

How so? The API in `build` has absolutely no isolation, and you are welcome to use it! The isolation is provided in a different module, `build.env`, which we use to setup an isolated environment before calling the API in `build`.

**Comment by gaborbernat** [comment]
@KOLANICH if you're against all the python packaging ecosystem mandates via PEPs (and from the sound of it you're), then probably better of not using tools that try to follow the standards. For what's worth pip -e is also not backed by any standard and as such has little guarantees at the moment.🤔

> IMHO all this build isolation brings no real benefits and has significant drawbacks

The python packaging standards doesn't agree with you on this per explanation in PEP 517 and 518. It's a must if you care about reproducible builds, otherwise the caller needs to ensure the environment invariants and that's very brittle. Can work, and hence we allow via the no isolation flag, but we don't recommend it as the go to default way.

**Comment by pganssle** [comment]
> That is a reasonable use-case.

I really don't think it is. Why use `build ` at all in the case of something that requires a development (!!) install of a setuptools plugin? `build` is for producing artifacts for distribution to other people. If you need a local install of something, use `pip`. How would someone install an sdist built this way?

> I think the forceful build isolation in PEP 517 is very opinionated.

I don't think opinionated is the right word here. It's done this way because without isolation you are likely to use very old versions of your packaging tools. These packaging tools are updated to keep up with changing standards and improvements to the ecosystem, and a lot of times people using very old tools have polluted PyPI with broken packages with broken metadata, which is a bad experience for end users.

I'm -0 (YAGNI) about passing this flag to `build ` via an environment variable. It's marginally better than a global flag, and it does solve the problem of "a tool wraps build", but I think in most cases the tool itself should be exposing this option. It's not like `pip`, where there are a lot of user-specific global options like specifying policies on firewalls or binaries.

TBH, it almost would make sense for the `no-isolation` command to be a separate command anyway, since it's used in a fundamentally different way.

**Comment by KOLANICH** [comment]
>The API in build has absolutely no isolation, and you are welcome to use it! 

Thanks for the info.

>if you're against all the python packaging ecosystem mandates via PEPs (and from the sound of it you're), then probably better of not using tools

hm... sounds like `then probably better of not using python`. For some ecosystems, like node.js I have really no choice but not to use them. But the world doesn't work like this, you know. I.e. I cannot stay away of node.js as long as M$ have chosen to use it fo GitHub CI. Advices like "not to use" are just impractical.

>The python packaging standards

are not dogmas. In fact there exist some harmful standards.

> It's a must if you care about reproducible builds

I both aggree and disaggree. It depends a lot on the way the packages used for building are written. I.e. they may use unpredictable stuff, and the result will be non-deterministic even if the versions are pinned. Or there may be different versions, but behaving identically. It is not the matter of version pinning, it is more a matter of contract defined and implemented by each tool, and features used by the buildee.

>contract dictates the build behaves by using isolated build environments. If you can override that you end up with all kind of weird failure and unsupported/unexpected behavior.

I am OK with that.

**Comment by FFY00** [comment]
> I really don't think it is. Why use `build ` at all in the case of something that requires a development (!!) install of a setuptools plugin? `build` is for producing artifacts for distribution to other people. If you need a local install of something, use `pip`. How would someone install an sdist built this way?

It seems to me that they are working on the setuptools plugins, hence the editable install. You should have a way to call the PEP 517 hooks, I think that is a reasonable use case. But that is besides the point, this is one of the cases where you want to bring your own environment, and as I have made clear in the past, I think that is valid, and the reason for `--no-isolation`.

> It's a must if you care about reproducible builds

Sorry, but it is absolutely not. No environment details get registered when you build a package, you cannot recreate the original environment without this information.

In fact, in Arch Linux packaging, which does not use the isolation described in PEP 517, we can achieve proper reproducible builds. This is because we save the environment information in the package metadata, so anyone can take the package, recreate the environment and build the exact same package, to the binary level -- you will have the same checksum.

What isolation does is improve reproducibility for general users, by controlling some aspects of the environment. But doing a build in isolation does not guarantee that you will be able to reproduce a package. So the PEP 517 isolation really helps the general public with reproducibility, but it is not a must for reproducible builds, it actually cannot achieve it.

**Comment by gaborbernat** [comment]
If you care about true reproducibility the standard allows you to pin all your build dependencies to an exact version. Or if you know your package builds with a range of versions you can use that range specifier to accept multiple versions.

**Comment by KOLANICH** [comment]
>I don't think opinionated is the right word here. It's done this way because without isolation you are likely to use very old versions of your packaging tools. 

I use setuptools from git. As I have said, pip is intentially set up in the way it won't install anything from pypi. So, if setuptools was locked to enforce build isolation, it would have failed to build anything on my machine.

>Why use build at all in the case of something that requires a development (!!) install of a setuptools plugin? build is for producing artifacts for distribution to other people. 

I don't separate development from production. For my use cases it is not needed. I just install own plugins using `-e` (in fact most of my own libs are installed this way) and then when in process of building some software it turns out that something is missing or broken in them (they are very immature and often need improvement), I just fix them.

**Comment by gaborbernat** [comment]
> I use setuptools from git.

This is very unusual. Though might be valid from your point of view and case. I feel this use case is something the generic Python Packaging ecosystem is not planning/intending to support, so tools that follow the Python packaging standard defined in PEPs you'll likely keep running into problems.

> I just install own plugins using `-e`

Technically this too would require installing build packages once https://discuss.python.org/t/specification-of-editable-installation/1564 becomes a thing. So I'm not sure your workflow will be supported in the mid-term.

> > contract dictates the build behaves by using isolated build environments. If you can override that you end up with all kind of weird failure and unsupported/unexpected behavior.
> 
> I am OK with that.


But as the tool developer, I'm personally not. It causes me endless bug reports from users that are less advanced and technically able. Tools, in general, are designed and aimed towards a wide audience and solve their problem as robust as possible.

In conclusion, I'm not saying that your use case is or is not valid. What I'm saying is that I don't think it's a problem for the wide audience and as such, I don't think we should support it out of the box. Python at the end of the day is a dynamic language, if you want to alter the per-contract defined behavior of tools (in which direction this discussion is going) you're free to do so via either installing patch pth files and monkeypatch the parts you want to alter. However, IMHO adding knobs to a tool that allows the user to break the contract of the tool easily via an environment variable or config file is something we should avoid.

**Comment by KOLANICH** [comment]
>In general, tools are designed and aimed towards a wide audience and solve their problem as robust as possible.

I don't think that the tools for programmers are aimed to "wide audience" of non-programmers. Ones decided to become programmers are not "general audience, they are the ones that are expected to know the tools they use enough.

>Technically this, too, would require installing build packages once https://discuss.python.org/t/specification-of-editable-installation/1564 becomes a thing.

I don't understand how the way pip `-e` works is related to this issue. Editable install must remain an editable install, right, no matter if a wheel is built or not for that. The problem with build isolation is that it tries to not to use system-installed packages. It is completely orthogonal to the way editable installs are done.

**Comment by gaborbernat** [comment]
>  Ones decided to become programmers are not "general audience, they are the ones that are expected to know the tools they use enough.

I wasn't comparing people against the general population of the world. I was comparing it against the general python programmer, which tends to have limited knowledge about python packaging in my experience (think data scientist, casual scriptwriters, etc.). Packaging regularly comes out in Python surveys as the hardest thing about Python so I'm personally on the side of let's make it as simple as possible.

**Comment by FFY00** [comment]
> If you care about true reproducibility the standard allows you to pin all your build dependencies to an exact version. Or if you know your package builds with a range of versions you can use that range specifier to accept multiple versions.

That makes reproducible build an opt-in from the upstream, instead of a feature for the users. And pinning dependencies like that is not usually good, as you'll be more likely to run into conflicts. I do not consider this a reproducible builds solution, it's just a mechanism to provide saner defaults to general users.

**Comment by gaborbernat** [comment]
> That makes reproducible build an opt-in from the upstream, instead of a feature for the users.

If the user can opt-in, isn't it a feature?

> And pinning dependencies like that is not usually good, as you'll be more likely to run into conflicts.

Why? If you pin to ``setuptools==44.0.0`` what will you be conflicting with? (ignoring the fact that upstream distributions may find it hard to satisfy that dependency).

**Comment by FFY00** [comment]
> If the user can opt-in, isn't it a feature?

The upstream needs to opt-in not the user.

> Why? If you pin to `setuptools==44.0.0` what will you be conflicting with? (ignoring the fact that upstream distributions may find it hard to satisfy that dependency).

What if you have to install other package that requires `setuptools==45.0.0`?

**Comment by gaborbernat** [comment]
If they also pin and you're clashing means you never had a valid build environment because both never can be satisfied at the same time.

**Comment by FFY00** [comment]
That is the issue.

> you'll be more likely to run into conflicts

Pinning dependencies when it is not required is bad because you will be more likely to have conflicts. IMO you shouldn't be doing it.

**Comment by pganssle** [comment]
> It seems to me that they are working on the setuptools plugins, hence the editable install. You should have a way to call the PEP 517 hooks, I think that is a reasonable use case. But that is besides the point, this is one of the cases where you want to bring your own environment, and as I have made clear in the past, I think that is valid, and the reason for `--no-isolation`.

Yeah, but if this is part of someone's test suite, it's only relevant for those test builds. I don't see that it's a big deal to include `build -n` in your `make test` target or whatever. There is no reason to have a global configuration for this.

> In fact, in Arch Linux packaging, which does not use the isolation described in PEP 517, we can achieve proper reproducible builds. This is because we save the environment information in the package metadata, so anyone can take the package, recreate the environment and build the exact same package, to the binary level -- you will have the same checksum.

To be clear, Arch Linux is using `--no-isolation` not because build isolation is not useful, but because `makepkg` provides a different form of build isolation (arguably better along many dimensions, with the one difference being that I believe the dependencies from `build` will leak into the environment) that is not compatible with `build`'s built-in isolation.

I agree that build isolation is not necessary or sufficient for reproducible builds, though to some extent Bernat is right in that the widespread use of build isolation would be a pretty significant improvement to the state of the current ecosystem with respect to reproducibility. It will catch most cases where the `pyproject.toml` doesn't accurately reflect what is needed to build the project. The other thing it does for us is minimize the degree to which people are producing packages with super old and buggy versions of their backends (`setuptools` particularly). There are many bad packages on PyPI that have bad metadata or other weird bugs because someone is building them with whatever version of `setuptools` they happen to have on their computer.

In any case — arguments about the extent to which build isolation is desirable to the side — I think we're all in agreement that this isn't something that should be configured globally, and as such there's no need for any sort of global or local config file here. @FFY00 can we close this?

**Comment by FFY00** [comment]
I am closing this, if you have some extra use-cases for us to consider, please re-open. Unfortunately the one presented is not strong enough to justify :confused:
